sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+            let mut value = Vec::<u8>::with_capacity(value_len as usize);
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: value_len as u64,
+                    is_negative: false,
+                }
+            };
+            unsafe {
+                sol_bignum_mod_inv(
+                    &lhs_ffi_in as *const _ as *const u64,
+                    &rhs_ffi_in as *const _ as *const u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    value,
+                }
+            }
+        }
+    }
+
+    /// Log a `BigNum` from a program to the solana logs
+    pub fn log(&self) {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            crate::program_stubs::sol_log(&self.value.to_dec_str().unwrap());
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_log(lhs_ffi_in_addr: *const u64) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            unsafe { sol_bignum_log(&lhs_ffi_in as *const _ as *const u64) };
+        }
+    }
+}
+
+impl PartialEq for BigNumber {
+    fn eq(&self, other: &Self) -> bool {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.value == other.value && self.value.is_negative() == other.value.is_negative()
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value == other.value && self.is_negative == other.is_negative
+        }
+    }
+}
+impl Clone for BigNumber {
+    fn clone(&self) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            match openssl::bn::BigNum::from_slice(&self.value.to_vec()) {
+                Ok(bn) => Self { value: bn },
+                Err(_) => panic!(),
+            }
+        }
+        #[cfg(target_arch = "bpf")]
+        Self {
+            value: self.value.clone(),
+            is_negative: self.is_negative,
+        }
+    }
+}
+
+impl Default for BigNumber {
+    fn default() -> Self {
+        BigNumber::new()
+    }
+}
+impl fmt::Debug for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            write!(f, "{}", self.value.to_dec_str()?)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+impl fmt::Display for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```

Comment:
Who uses this?

Reply:
@jackcmay For debugging we use `println!` in non-BPF and `msg!()` in BPF. The way we have conditional constructs of the `BigNumber` struct I can't use the Traits at the higher level so I have to implement and can't just use the macro.

Reply:
Hmm, because of the conditional you can't use `#[derive(Debug, Display)]?

Reply:
@jackcmay Aye, error posted above